### PR TITLE
Fix early disposal of the IApiResponse instance

### PR DESCRIPTION
**Pull request by kevingosse**
IApiResponse was disposed before reading the actual response.

I separated the retry logic and the actual sending operation into separate methods, because the workflow was becoming too hard to follow (that explains why neither myself or any reviewer noticed the mistake).

I also noticed that the template for the logs was wrong, so I fixed that as well.

### CODE REVIEW

File: src/Datadog.Trace/Agent/Api.cs [hunk 1]
```diff
@@ -74,72 +74,41 @@ public async Task<bool> SendTracesAsync(Span[][] traces)
...
-                    }
+                    success = await SendTracesAsync(traces, request).ConfigureAwait(false);
                 }
                 catch (Exception ex)
                 {
 #if DEBUG
                     if (ex.InnerException is InvalidOperationException ioe)
                     {
-                        Log.Error("An error occurred while sending traces to the agent at {Endpoint}\n{Exception}", ex, _tracesEndpoint, ex.ToString());
+                        Log.Error(ex, "An error occurred while sending traces to the agent at {0}\n{1}", _tracesEndpoint, ex.ToString());
```
Comment (colin-higgins) [inline src/Datadog.Trace/Agent/Api.cs:89]: Maybe `Environment.NewLine` instead of `\n`?

Comment (kevingosse) [reply src/Datadog.Trace/Agent/Api.cs:89]: Well actually we shouldn't even be concatenating the exception since we pass it using the dedicated overload. Gonna fix this

File: src/Datadog.Trace/Agent/Api.cs [hunk 2]
```diff
@@ -195,6 +146,63 @@ private static HashSet<ulong> GetUniqueTraceIds(Span[][] traces)
...
+                {
+                    // don't bother creating the tags array if trace metrics are disabled
+                    string[] tags = { $"status:{response.StatusCode}" };
+
+                    // count every response, grouped by status code
+                    _statsd.AppendIncrementCount(TracerMetricNames.Api.Responses, tags: tags);
+                }
+
+                // Attempt a retry if the status code is not SUCCESS
+                if (response.StatusCode < 200 || response.StatusCode > 300)
```
Comment (lucaspimentel) [inline src/Datadog.Trace/Agent/Api.cs:179]: Only 2xx is success, not 300.

```suggestion
                if (response.StatusCode < 200 || response.StatusCode >= 300)
```

File: src/Datadog.Trace/Agent/Api.cs [hunk 3]
```diff
@@ -195,6 +146,63 @@ private static HashSet<ulong> GetUniqueTraceIds(Span[][] traces)
...
+                try
+                {
+                    _statsd?.AppendIncrementCount(TracerMetricNames.Api.Requests);
+                    response = await request.PostAsync(traces, _formatterResolver).ConfigureAwait(false);
+                }
+                catch
+                {
+                    // count the exceptions thrown by the HttpClient,
+                    // not responses with 5xx status codes
+                    // (which cause EnsureSuccessStatusCode() to throw below)
```
Comment (lucaspimentel) [inline src/Datadog.Trace/Agent/Api.cs:164]: Nit: This comment is outdated, we can't call `EnsureSuccessStatusCode()` anymore.

Comment (lucaspimentel) [reply src/Datadog.Trace/Agent/Api.cs:164]: Actually, I think `WebRequest` throws for 5xx response codes, so we're probably not counting the API errors correctly anymore, but that's out of scope for this PR.

File: src/Datadog.Trace/Agent/Api.cs (hunk 2 above)
Comment (kevingosse) [reply src/Datadog.Trace/Agent/Api.cs:179]: You're right, but that's a copy/paste of the old code. Let's fix that in a separate PR

**Comment by ralmsdeveloper** [comment]
Hello @zacharycmontoya,

Is this a new feature or a bug fix?
I updated my entire environment yesterday, more than 30 applications with version 1.9.0, I would like to know the negative impact that I will have!

**Comment by zacharycmontoya** [comment]
Hi @ralmsdeveloper , this is a bug fix so if you have time please do update to 1.19.1! Your traces will still be sent to the Datadog agent but the .NET Tracer will not read the response from the agent that may request an update to sampling rates and this will also add extra noise to the Tracer log file complaining about an `ObjectDisposedException`, but that is the extent of the issue.

**Comment by ralmsdeveloper** [comment]
@zacharycmontoya Thank u, I will set up a new task force with the engineering team.
